aea/agent_loop.py

```diff
@@ -0,0 +1,406 @@
+    return value
+
+
+class AsyncState:
+    """Awaitable state."""
+
+    def __init__(self, initial_state: Any = None, loop: AbstractEventLoop = None):
+        """Init async state.
+
+        :param initial_state: state to set on start.
+        :param loop: optional asyncio event loop.
+        """
+        self._state = initial_state
+        self._watchers: Set[Future] = set()
+        self._loop = loop or asyncio.get_event_loop()
+
+    @property
+    def state(self) -> Any:
+        """Return current state."""
+        return self._state
+
+    @state.setter
+    def state(self, state: Any) -> None:
+        """Set state."""
+        if self._state == state:
+            return
+        self._state_changed(state)
+        self._state = state
+
+    def _state_changed(self, state: Any) -> None:
+        """Fulfill watchers for state."""
+        for watcher in list(self._watchers):
+            if state in watcher._states:  # type: ignore
+                self._loop.call_soon_threadsafe(
+                    watcher.set_result, (self._state, state)
+                )
+
+    async def wait(self, state_or_states: Union[Any, List[Any]]) -> Tuple[Any, Any]:
+        """Wait state to be set.
+
+        :params state_or_states: state or list of states.
+        :return: tuple of previous state and new state.
+        """
+        states = ensure_list(state_or_states)
+
+        if self._state in states:
+            return (None, self._state)
+
+        watcher: Future = Future()
+        watcher._states = states  # type: ignore
+        self._watchers.add(watcher)
+        try:
+            return await watcher
+        finally:
+            self._watchers.remove(watcher)
+
+
+class PeriodicCaller:
+    """Schedule a periodic call of callable using event loop."""
+
+    def __init__(
+        self,
+        callback: Callable,
+        period: float,
+        start_at: Optional[datetime.datetime] = None,
+        exception_callback: Optional[Callable] = None,
+        loop: Optional[AbstractEventLoop] = None,
+    ):
+        """
+        Init periodic caller.
+
+        :param callback: function to call periodically
+        :param period: period in seconds.
+        :param start_at: optional first call datetime
+        :param exception_callback: optional handler to call on exception raised.
+        :param loop: optional asyncio event loop
+        """
+        self._loop = loop or asyncio.get_event_loop()
+        self._periodic_callable = callback
+        self._start_at = start_at or datetime.datetime.now()
+        self._period = period
+        self._timerhandle: Optional[TimerHandle] = None
+        self._exception_callback = exception_callback
+
+    def _callback(self) -> None:
+        """Call on each schduled call."""
+        self._schedule_call()
+        try:
+            self._periodic_callable()
+        except Exception as exception:
+            if not self._exception_callback:
+                raise
```

Comment:
is it a good use case for `raise from exception`?

Reply:
it reraises the last handled exception.
https://docs.python.org/3/tutorial/errors.html

Reply:
yes, but I was thinking that we could give some more information to the user on what happened: which was the periodic task, etc.
In other words, to enrich the exception, but raising from the same exception context so the stack trace won't be lost.

Anyway, not a very crucial feature to have at the moment. It was just a thought.

Reply:
but it's already done, cause it's reraised in case no exception_callback provided, in exception callback there is a logging and agent termination code